# Statistical Analysis Overview

In this analysis, I perform several key statistical evaluations to understand the dataset deeply before modeling:

1. **Descriptive Statistics:**  
   - Summarize basic statistics like mean, median, standard deviation, min, max, and quartiles for each feature.

2. **Skewness Analysis:**  
   - Measure asymmetry in the distribution of each variable.  
   - Identify whether features are right-skewed (long right tail), left-skewed (long left tail), or symmetrical, which helps understand the data distribution and potential outliers.

3. **Kurtosis Analysis:**  
   - Assess the "tailedness" of each feature’s distribution.  
   - High kurtosis indicates heavy tails or outliers; low kurtosis suggests light tails.

4. **Correlation Analysis:**  
   - Examine linear relationships between pairs of features.  
   - Identify highly correlated variables (positive or negative) to avoid redundancy in modeling and multicollinearity issues.

This step-by-step statistical analysis helps to:  
- Detect data imbalances and anomalies,  
- Select relevant features for modeling,  
- Understand the underlying physical or operational system reflected in the data.

In [10]:
!pip install kagglehub --quiet

import kagglehub
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.tsa.stattools import adfuller

In [11]:
# Download the dataset
path = kagglehub.dataset_download("anshtanwar/metro-train-dataset")
print("Dataset path:", path)

Dataset path: /root/.cache/kagglehub/datasets/anshtanwar/metro-train-dataset/versions/1


In [12]:
dataset_path = "/root/.cache/kagglehub/datasets/anshtanwar/metro-train-dataset/versions/1"
print("Files in dataset folder:")
print(os.listdir(dataset_path))


Files in dataset folder:
['MetroPT3(AirCompressor).csv']


In [13]:
file_path = os.path.join(dataset_path, "MetroPT3(AirCompressor).csv")
df = pd.read_csv(file_path)

In [14]:
# Preview
df.head()

Unnamed: 0            timestamp    TP2    TP3     H1  DV_pressure  \
0           0  2020-02-01 00:00:00 -0.012  9.358  9.340       -0.024   
1          10  2020-02-01 00:00:10 -0.014  9.348  9.332       -0.022   
2          20  2020-02-01 00:00:19 -0.012  9.338  9.322       -0.022   
3          30  2020-02-01 00:00:29 -0.012  9.328  9.312       -0.022   
4          40  2020-02-01 00:00:39 -0.012  9.318  9.302       -0.022   

   Reservoirs  Oil_temperature  Motor_current  COMP  DV_eletric  Towers  MPG  \
0       9.358           53.600         0.0400   1.0         0.0     1.0  1.0   
1       9.348           53.675         0.0400   1.0         0.0     1.0  1.0   
2       9.338           53.600         0.0425   1.0         0.0     1.0  1.0   
3       9.328           53.425         0.0400   1.0         0.0     1.0  1.0   
4       9.318           53.475         0.0400   1.0         0.0     1.0  1.0   

   LPS  Pressure_switch  Oil_level  Caudal_impulses  
0  0.0              1.0        1.0              1.0  
1  0.0              1.0        1.0              1.0  
2  0.0              1.0        1.0              1.0  
3  0.0              1.0        1.0              1.0  
4  0.0              1.0        1.0              1.0

In [15]:
df.describe()

Unnamed: 0           TP2           TP3            H1   DV_pressure  \
count  1.516948e+06  1.516948e+06  1.516948e+06  1.516948e+06  1.516948e+06   
mean   7.584735e+06  1.367826e+00  8.984611e+00  7.568155e+00  5.595619e-02   
std    4.379053e+06  3.250930e+00  6.390951e-01  3.333200e+00  3.824015e-01   
min    0.000000e+00 -3.200000e-02  7.300000e-01 -3.600000e-02 -3.200000e-02   
25%    3.792368e+06 -1.400000e-02  8.492000e+00  8.254000e+00 -2.200000e-02   
50%    7.584735e+06 -1.200000e-02  8.960000e+00  8.784000e+00 -2.000000e-02   
75%    1.137710e+07 -1.000000e-02  9.492000e+00  9.374000e+00 -1.800000e-02   
max    1.516947e+07  1.067600e+01  1.030200e+01  1.028800e+01  9.844000e+00   

         Reservoirs  Oil_temperature  Motor_current          COMP  \
count  1.516948e+06     1.516948e+06   1.516948e+06  1.516948e+06   
mean   8.985233e+00     6.264418e+01   2.050171e+00  8.369568e-01   
std    6.383070e-01     6.516261e+00   2.302053e+00  3.694052e-01   
min    7.120000e-01     1.540000e+01   2.000000e-02  0.000000e+00   
25%    8.494000e+00     5.777500e+01   4.000000e-02  1.000000e+00   
50%    8.960000e+00     6.270000e+01   4.500000e-02  1.000000e+00   
75%    9.492000e+00     6.725000e+01   3.807500e+00  1.000000e+00   
max    1.030000e+01     8.905000e+01   9.295000e+00  1.000000e+00   

         DV_eletric        Towers           MPG           LPS  \
count  1.516948e+06  1.516948e+06  1.516948e+06  1.516948e+06   
mean   1.606106e-01  9.198483e-01  8.326640e-01  3.420025e-03   
std    3.671716e-01  2.715280e-01  3.732757e-01  5.838091e-02   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
50%    0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
75%    0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
max    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   

       Pressure_switch     Oil_level  Caudal_impulses  
count     1.516948e+06  1.516948e+06     1.516948e+06  
mean      9.914368e-01  9.041556e-01     9.371066e-01  
std       9.214078e-02  2.943779e-01     2.427712e-01  
min       0.000000e+00  0.000000e+00     0.000000e+00  
25%       1.000000e+00  1.000000e+00     1.000000e+00  
50%       1.000000e+00  1.000000e+00     1.000000e+00  
75%       1.000000e+00  1.000000e+00     1.000000e+00  
max       1.000000e+00  1.000000e+00     1.000000e+00

In [16]:
df.isnull().sum()

Unnamed: 0         0
timestamp          0
TP2                0
TP3                0
H1                 0
DV_pressure        0
Reservoirs         0
Oil_temperature    0
Motor_current      0
COMP               0
DV_eletric         0
Towers             0
MPG                0
LPS                0
Pressure_switch    0
Oil_level          0
Caudal_impulses    0
dtype: int64

In [17]:
df['timestamp'] = pd.to_datetime(df['timestamp'])
df.set_index('timestamp', inplace=True)

In [18]:
df.drop(columns=["Unnamed: 0"], inplace=True)

In [19]:
df.describe()

TP2           TP3            H1   DV_pressure    Reservoirs  \
count  1.516948e+06  1.516948e+06  1.516948e+06  1.516948e+06  1.516948e+06   
mean   1.367826e+00  8.984611e+00  7.568155e+00  5.595619e-02  8.985233e+00   
std    3.250930e+00  6.390951e-01  3.333200e+00  3.824015e-01  6.383070e-01   
min   -3.200000e-02  7.300000e-01 -3.600000e-02 -3.200000e-02  7.120000e-01   
25%   -1.400000e-02  8.492000e+00  8.254000e+00 -2.200000e-02  8.494000e+00   
50%   -1.200000e-02  8.960000e+00  8.784000e+00 -2.000000e-02  8.960000e+00   
75%   -1.000000e-02  9.492000e+00  9.374000e+00 -1.800000e-02  9.492000e+00   
max    1.067600e+01  1.030200e+01  1.028800e+01  9.844000e+00  1.030000e+01   

       Oil_temperature  Motor_current          COMP    DV_eletric  \
count     1.516948e+06   1.516948e+06  1.516948e+06  1.516948e+06   
mean      6.264418e+01   2.050171e+00  8.369568e-01  1.606106e-01   
std       6.516261e+00   2.302053e+00  3.694052e-01  3.671716e-01   
min       1.540000e+01   2.000000e-02  0.000000e+00  0.000000e+00   
25%       5.777500e+01   4.000000e-02  1.000000e+00  0.000000e+00   
50%       6.270000e+01   4.500000e-02  1.000000e+00  0.000000e+00   
75%       6.725000e+01   3.807500e+00  1.000000e+00  0.000000e+00   
max       8.905000e+01   9.295000e+00  1.000000e+00  1.000000e+00   

             Towers           MPG           LPS  Pressure_switch  \
count  1.516948e+06  1.516948e+06  1.516948e+06     1.516948e+06   
mean   9.198483e-01  8.326640e-01  3.420025e-03     9.914368e-01   
std    2.715280e-01  3.732757e-01  5.838091e-02     9.214078e-02   
min    0.000000e+00  0.000000e+00  0.000000e+00     0.000000e+00   
25%    1.000000e+00  1.000000e+00  0.000000e+00     1.000000e+00   
50%    1.000000e+00  1.000000e+00  0.000000e+00     1.000000e+00   
75%    1.000000e+00  1.000000e+00  0.000000e+00     1.000000e+00   
max    1.000000e+00  1.000000e+00  1.000000e+00     1.000000e+00   

          Oil_level  Caudal_impulses  
count  1.516948e+06     1.516948e+06  
mean   9.041556e-01     9.371066e-01  
std    2.943779e-01     2.427712e-01  
min    0.000000e+00     0.000000e+00  
25%    1.000000e+00     1.000000e+00  
50%    1.000000e+00     1.000000e+00  
75%    1.000000e+00     1.000000e+00  
max    1.000000e+00     1.000000e+00

### 🧠 My Understanding of the Dataset

After looking at the statistical summary of the MetroPT-3 compressor dataset, here’s how I understand each feature and what’s going on in the system:

#### 🔵 Pressure-Related Sensors:
- **TP2**: This seems to measure compressor pressure. The average is ~ 1.37 bar, but the standard deviation is high (~3.25), and I noticed some negative values – maybe due to sensor issues or when the compressor is off. A large portion of readings is close to zero.

- **TP3**: It looks like a stable pneumatic panel pressure sensor. The mean is ~8.98 bar and the values are tightly clustered – this indicates normal and stable system operation.

- **H1**: Measures pressure from the cyclonic separator. It shows high variation and even some negative values (not expected), which I think needs further investigation.


- **DV_pressure**: Seems to measure the pressure drop at the air dryer. It has a very low mean (0.056), and the median is actually negative. Most of the time the system seems offloaded here.

- **Reservoirs**: Almost identical to TP3, which makes sense because it’s likely drawing from the same pressure line.

#### 🌡 Operational Parameters:
- **Oil_temperature**: On average, it's ~62.6°C which is reasonable. The max is 89°C – this might indicate rare overheating. Something to keep an eye on.
- **Motor_current**: This tells me a lot! The current goes from 0A (off) to 4A (offloaded), 7A (loaded), and up to 9A during startup. The mean is around 2.05A, but it's not a good central measure since it's a mix of many states.

#### ⚙ Digital/Binary Features (1 = Active, 0 = Inactive):
- **COMP**: The intake valve is active about 84% of the time – that looks healthy.

- **DV_eletric**: This valve seems to do the opposite of COMP – it’s only active ~16% of the time.

- **TOWERS**: These are air dryer towers – active 92% of the time. It looks like Tower 2 is used more often.

- **MPG**: Load starter – behavior is quite similar to COMP.

- **LPS (Low Pressure Switch)**: Only active in rare cases (~0.34%). This means the system pressure is mostly stable.

- **Pressure_switch**: Almost always on (99.1%) – which is expected for a healthy system.

- **Oil_level**: Around 90% of readings say oil level is OK. That’s a good sign.


- **Caudal_impulses**: These signals seem to track airflow pulses. Active ~93.7% of the time, showing consistent flow.

### 🔍 Key Patterns I Observed:
- TP3 and Reservoir pressures are very stable → good sign.
- TP2 and H1 show odd/negative values → maybe off states or sensor error.
- Motor current helps to identify different operational phases clearly.
- COMP and MPG activate together; DV_eletric is kind of their opposite.
- Rare LPS activation means pressure rarely drops dangerously.
- Oil levels and pressure switches are healthy 90–99% of the time.

### ✅ What I Plan to Do Next:TODO

1. **Clean** up those strange negative values (TP2, H1, DV_pressure).

2. **Engineer** some useful features:
   - Add labels for motor state (Off, Offloaded, Loaded, Startup).
   - Compute pressure differentials like `TP3 - TP2`.
   - Track how long COMP or DV_eletric stays ON or OFF.
3. **Detect anomalies**:
   - Flag when LPS activates.
   - Catch high oil temperature events.
   - Watch for pressure switch turning off unexpectedly.




### ✅  **Skewness and Kurtosis**

* **Purpose**: Understand distribution shape.

  * **Skewness** tells if the data is symmetric.
  * **Kurtosis** indicates whether the data is heavy/light-tailed compared to a normal distribution.


In [21]:
df.skew(numeric_only=True)

TP2                 1.986279
TP3                -0.877349
H1                 -1.764960
DV_pressure         5.720572
Reservoirs         -0.886154
Oil_temperature    -0.045791
Motor_current       0.448561
COMP               -1.824324
DV_eletric          1.848671
Towers             -3.092490
MPG                -1.782406
LPS                17.011747
Pressure_switch   -10.667099
Oil_level          -2.745832
Caudal_impulses    -3.600981
dtype: float64


**My Understanding of Feature Skewness (Asymmetry)**

I used **skewness** to understand how the values of each feature are distributed. Here's what I know:

* A **skewness of 0** means a feature is symmetric like a normal distribution.
* A **positive skew** tells me the values are mostly small, with some rare large ones (right-tailed).
* A **negative skew** means the values are mostly high, with rare small ones (left-tailed).

### 🔍 **What I Learned From Each Feature**

#### ✅ **1. Right-Skewed Features (Positive Values)**

* **DV\_pressure (5.72)**:
  This tells me the pressure is mostly low or zero, and only spikes occasionally. That matches the behavior during offloaded states—mostly off, but some sharp pressure drops when the compressor loads.

* **LPS (17.01)**:
  Super right-skewed. I see it’s almost always off (0), and only rarely turns on. Makes sense, since this is only triggered during emergency low-pressure situations.

* **DV\_eletric (1.85)**:
  It stays off most of the time, only switching on when the compressor is loaded. That fits with what I expect from the system logic.

* **TP2 (1.99)**:
  Shows low pressure readings usually, and occasional spikes. Probably reflects inlet pressure during idle or start.

#### ✅ **2. Left-Skewed Features (Negative Values)**

* **Towers (-3.09)**:
  This tells me one tower (Tower 2) is mostly in use. The value is usually high (1), with rare switches.

* **Pressure\_switch (-10.67)**:
  It's nearly always ON (1), which makes sense because it should be active during most operations. When it drops to 0, it might be a signal of a shutdown or fault.

* **COMP (-1.82)** and **MPG (-1.78)**:
  These features are mostly active, which shows that the compressor and intake valve are usually engaged. That matches my understanding of a running machine.

* **Oil\_level (-2.75)**:
  Mostly 1, meaning oil is usually at a good level. When it’s 0, it might mean a rare low-oil condition — could be a useful fault indicator.

* **H1 (-1.76)**:
  Tells me discharge pressure is usually high, which I expect in normal running mode.

* **Caudal\_impulses (-3.60)**:
  This is almost always active (pulsing = 1). That fits the idea that airflow is continuous when the system is on.

* **TP3 (-0.88)** and **Reservoirs (-0.89)**:
  Both are mostly high-pressure states. Tells me the system maintains pressure well.

#### ✅ **3. Nearly Symmetric Features (Close to 0)**

* **Oil\_temperature (-0.05)**:
  This one really caught my eye — it's almost perfectly symmetrical. That means the temperature is stable, and the system has good thermal regulation.

* **Motor\_current (0.45)**:
  Slightly right-skewed, meaning there are more low-current readings. Makes sense, since the motor is off or in offload mode a lot, and current spikes only during load.


### 💡 **How I Interpreted It All**

* **Digital/Binary Features** like `COMP`, `DV_eletric`, `LPS`, and `Towers` have extreme skewness. That’s expected — they spend most of their time in one state.

* **Pressure Features** help me see the real system behavior:

  * `TP2` is usually low (before compression), while `TP3` and `Reservoirs` stay high (after compression).
  * `DV_pressure` gives pressure drop — mostly zero, but spikes under load.

* **Red Flags**:
  If I ever see 0s in left-skewed features like `Pressure_switch` or `Oil_level`, I’d suspect an anomaly.
  If I see huge values in right-skewed features like `DV_pressure`, it could be a pressure surge or emergency event.



In [22]:
df.kurtosis(numeric_only=True)

TP2                  2.060720
TP3                  7.395316
H1                   1.296244
DV_pressure         38.714964
Reservoirs           7.482164
Oil_temperature     -0.057793
Motor_current       -1.450753
COMP                 1.328158
DV_eletric           1.417588
Towers               7.563502
MPG                  1.176973
LPS                287.399911
Pressure_switch    111.787147
Oil_level            5.539602
Caudal_impulses     10.967077
dtype: float64


**My Understanding of Feature Kurtosis**

I used **kurtosis** to understand how "peaky" or "flat" each feature’s distribution is. It shows how often extreme values (outliers) occur compared to a normal distribution.

Here’s how I interpret it:

* **Kurtosis ≈ 0** → Normal distribution (moderate tails)
* **Kurtosis > 3** → Leptokurtic → Very sharp peaks + heavy tails → more outliers
* **Kurtosis < 3** → Platykurtic → Flatter distribution → fewer outliers

Let me go feature by feature:


### 🔺 **Very High Kurtosis (Heavy Tails & Outliers)**

These features have **extreme outliers**, meaning the data is mostly near a central value but sometimes jumps far off:

* **LPS (287.40)**:
  This one really stood out. It’s binary but with huge kurtosis. To me, this means it’s almost always 0, but those rare 1s are statistically massive outliers. Matches emergency behavior.

* **Pressure\_switch (111.79)**:
  Again, mostly stuck at 1, but rare 0s are treated like deep statistical outliers. Feels like it flags critical system state changes.

* **DV\_pressure (38.71)**:
  Most of the time it’s zero, but when the system loads, pressure surges sharply. These occasional big values shoot the kurtosis way up.

* **Towers (7.56)**:
  Tower 2 (value = 1) is usually in use. When it switches to Tower 1 (0), it’s such a rare event that it gets treated like an outlier.

* **Reservoirs (7.48)**:
  Values stay tightly packed near high pressure, but there are occasional small dips or spikes, which create tail heaviness.

* **Caudal\_impulses (10.97)**:
  I noticed it pulses constantly, but rare 0s exist. These rare events have strong impact in stats, thus high kurtosis.

* **Oil\_level (5.54)**:
  The oil level stays normal (1) almost all the time. Rare drops to 0 are sharp and significant — and get counted as extreme.


### 🟡 **Moderately High Kurtosis (Mild Outliers)**

* **TP3 (7.39)**:
  It generally maintains a consistent high pressure but sometimes sees pressure drops or spikes. Those extremes give it high kurtosis.

* **COMP (1.33)**, **DV\_eletric (1.42)**, **MPG (1.18)**, **H1 (1.30)**:
  These have some mild peakedness. They usually sit at a dominant state but have enough variation for the tails to be just a bit heavier than normal. So not alarming — just shows some variability.


### ✅ **Nearly Normal Kurtosis (\~0)**

* **TP2 (2.06)**:
  Close to normal range. It fluctuates a bit but not many extreme values. I think this inlet pressure behaves more “normally.”

* **Oil\_temperature (-0.06)**:
  Very close to zero — I’d call it a perfect example of stable operation. Values are balanced without many outliers.


### 🔻 **Low Kurtosis (Flat or Wide Distributions)**

* **Motor\_current (-1.45)**:
  This surprised me. It shows a relatively **flat** distribution. So instead of being focused around a central mode, current spreads out more evenly across the range. Makes sense, as the motor might run at different levels — off, idle, or full load.


* Features like **LPS**, **Pressure\_switch**, and **DV\_pressure** act like on/off switches that are “quiet” most of the time but send a clear statistical shock when triggered.

* High kurtosis makes me alert for **rare but critical events** — like **oil level drops**, **pressure dips**, or **compressor switching states**.

* Flat kurtosis like in **motor current** tells me the system has many valid operation levels — not just one dominant value.


In [24]:
# Pearson for linear
pearson_corr = df.corr(method='pearson')
print(pearson_corr)

                      TP2       TP3        H1  DV_pressure  Reservoirs  \
TP2              1.000000 -0.011161 -0.961269     0.415025   -0.012403   
TP3             -0.011161  1.000000  0.224867    -0.153074    0.999993   
H1              -0.961269  0.224867  1.000000    -0.425513    0.226037   
DV_pressure      0.415025 -0.153074 -0.425513     1.000000   -0.153080   
Reservoirs      -0.012403  0.999993  0.226037    -0.153080    1.000000   
Oil_temperature  0.250710  0.401616 -0.161810     0.339697    0.401647   
Motor_current    0.697480  0.413756 -0.600178     0.302160    0.412691   
COMP            -0.955521  0.103295  0.971419    -0.423992    0.104509   
DV_eletric       0.947396 -0.078428 -0.958663     0.427813   -0.079640   
Towers          -0.616405  0.064937  0.628964    -0.285256    0.065729   
MPG             -0.941250  0.088343  0.954307    -0.417284    0.089555   
LPS              0.057651 -0.324556 -0.133178     0.011928   -0.325235   
Pressure_switch -0.069532  0.025240  0


✅ **Strong linear relationships** (good candidates for multicollinearity check):

* `TP3` and `Reservoirs` → **0.99999** → Almost perfect positive correlation.
* `H1` and `COMP` → **0.97**, `MPG` and `COMP` → **0.98** → Very strong positive linear relations.
* `H1` and `TP2` → **-0.96**, `DV_eletric` and `COMP` → **-0.96** → Strong **negative** linear relations.

⚠️ **High correlations (positive or negative)** may cause multicollinearity issues in linear models (like Linear Regression). Variables like `TP3`, `Reservoirs`, `H1`, `COMP`, and `MPG` are highly correlated — **you may consider dropping or combining them**.


✅ **Moderate linear relations**:

* `Motor_current` shows moderate correlation with `TP2`, `TP3`, `DV_eletric`.
* `Oil_temperature` has weak-moderate relations with `TP2`, `TP3`, `DV_pressure`.


❌ **Weak or negligible relationships**:

* Variables like `Caudal_impulses`, `LPS`, `Oil_level`, and `Pressure_switch` have **low correlation** with most other variables — likely **independent**, and **may be useful** if they have predictive power.


In [26]:
# Spearman for monotonic relationships (better for ranked or binary features)
spearman_corr = df.corr(method='spearman')
print(spearman_corr)

                      TP2       TP3        H1  DV_pressure  Reservoirs  \
TP2              1.000000  0.004374 -0.390699     0.550381    0.004138   
TP3              0.004374  1.000000  0.727916    -0.058441    0.999992   
H1              -0.390699  0.727916  1.000000    -0.149464    0.728673   
DV_pressure      0.550381 -0.058441 -0.149464     1.000000   -0.057275   
Reservoirs       0.004138  0.999992  0.728673    -0.057275    1.000000   
Oil_temperature  0.357251  0.433453  0.253255     0.437866    0.433640   
Motor_current    0.357963  0.453106  0.007653     0.024379    0.452028   
COMP            -0.626228  0.073643  0.629679    -0.257932    0.074764   
DV_eletric       0.612263 -0.046048 -0.603531     0.242463   -0.047153   
Towers          -0.404611  0.044187  0.410601    -0.154550    0.044901   
MPG             -0.607426  0.058269  0.607554    -0.245701    0.059375   
LPS              0.054337 -0.101087 -0.080080     0.017091   -0.101088   
Pressure_switch -0.047226  0.033739  0

In [28]:
zero_ratio = (df == 0).sum() / len(df)
print(zero_ratio)
constant_features = df.nunique() == 1
print(constant_features)

TP2                0.000074
TP3                0.000000
H1                 0.000253
DV_pressure        0.000262
Reservoirs         0.000000
Oil_temperature    0.000000
Motor_current      0.000000
COMP               0.163043
DV_eletric         0.839389
Towers             0.080152
MPG                0.167336
LPS                0.996580
Pressure_switch    0.008563
Oil_level          0.095844
Caudal_impulses    0.062893
dtype: float64
TP2                False
TP3                False
H1                 False
DV_pressure        False
Reservoirs         False
Oil_temperature    False
Motor_current      False
COMP               False
DV_eletric         False
Towers             False
MPG                False
LPS                False
Pressure_switch    False
Oil_level          False
Caudal_impulses    False
dtype: bool


In [30]:
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1

outliers = ((df < (Q1 - 1.5 * IQR)) | (df > (Q3 + 1.5 * IQR))).sum()
print(outliers)

TP2                269592
TP3                  4518
H1                 241327
DV_pressure        113111
Reservoirs           4526
Oil_temperature      5194
Motor_current           0
COMP               247328
DV_eletric         243638
Towers             121586
MPG                253840
LPS                  5188
Pressure_switch     12990
Oil_level          145391
Caudal_impulses     95406
dtype: int64



This shows how many outliers were detected in each feature using the Interquartile Range (IQR) method:

- Features with **many outliers** (large numbers) have a lot of extreme values beyond the typical range.  
  - Examples: `TP2 (269,592)`, `H1 (241,327)`, `COMP (247,328)`, `DV_eletric (243,638)`, `MPG (253,840)` — means these features have many unusual or extreme data points.

- Features with **few or no outliers** have mostly normal data within expected ranges.  
  - Examples: `Motor_current (0)`, `Reservoirs (4,526)`, `TP3 (4,518)` — means their data is more consistent without many extreme values.


Outliers can affect model performance and analysis, so knowing how many exist helps decide if we need to clean or handle those values carefully.

### ✅ **Coefficient of Variation (CV)**

* `CV = std / mean`
* Helps in comparing variability between features regardless of units.

In [31]:
cv = df.std(numeric_only=True) / df.mean(numeric_only=True)
print(cv)

TP2                 2.376713
TP3                 0.071132
H1                  0.440424
DV_pressure         6.833945
Reservoirs          0.071040
Oil_temperature     0.104020
Motor_current       1.122859
COMP                0.441367
DV_eletric          2.286097
Towers              0.295188
MPG                 0.448291
LPS                17.070317
Pressure_switch     0.092937
Oil_level           0.325583
Caudal_impulses     0.259065
dtype: float64


The CV measures relative variability by comparing the standard deviation to the mean for each feature:

- **High CV (>1)** indicates high variability relative to the mean, suggesting less consistency or more dispersion in data.  
  - Examples: `DV_pressure (6.83)`, `LPS (17.07)`, `DV_eletric (2.29)`, `TP2 (2.38)`.

- **Moderate CV (~0.3 to 1)** reflects moderate variability.  
  - Examples: `Motor_current (1.12)`, `COMP (0.44)`, `MPG (0.45)`, `H1 (0.44)`.

- **Low CV (<0.3)** indicates stable features with low relative variability.  
  - Examples: `TP3 (0.07)`, `Reservoirs (0.07)`, `Pressure_switch (0.09)`, `Oil_temperature (0.10)`, `Oil_level (0.33)`, `Caudal_impulses (0.26)`.

These CV values help identify which features have more relative dispersion, important for feature scaling, normalization, and understanding data consistency.


### ✅  **Transition Count**

* Count how many times a signal changes value (e.g., motor turning ON/OFF).
* Shows activity frequency or system switching behavior.


In [32]:
def count_transitions(series):
    return (series.diff() != 0).sum()

transition_counts = df.select_dtypes(include='number').apply(count_transitions)
print(transition_counts)

TP2                 803293
TP3                1460709
H1                 1372438
DV_pressure         611669
Reservoirs         1460884
Oil_temperature    1335225
Motor_current       935448
COMP                 33847
DV_eletric           31449
Towers               64581
MPG                  31013
LPS                    255
Pressure_switch      23629
Oil_level            15337
Caudal_impulses      15293
dtype: int64
